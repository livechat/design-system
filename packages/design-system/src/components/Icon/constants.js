export const IconSizeName = {
  XSmall: 'xsmall',
  Small: 'small',
  Medium: 'medium',
  Large: 'large',
  XLarge: 'xlarge'
};

export const IconSize = {
  [IconSizeName.XSmall]: {
    width: 16,
    height: 16
  },
  [IconSizeName.Small]: {
    width: 20,
    height: 20
  },
  [IconSizeName.Medium]: {
    width: 24,
    height: 24
  },
  [IconSizeName.Large]: {
    width: 32,
    height: 32
  },
  [IconSizeName.XLarge]: {
    width: 48,
    height: 48
  }
};

export const IconTypeName = {
  Primary: 'primary',
  Inverted: 'inverted',
  Link: 'link',
  Success: 'success',
  Warning: 'warning',
  Error: 'error'
};

export const IconColorMapper = {
  [IconTypeName.Primary]: '#424d57',
  [IconTypeName.Inverted]: '#fff',
  [IconTypeName.Link]: '#4384f5',
  [IconTypeName.Success]: '#4bb678',
  [IconTypeName.Warning]: '#efa842',
  [IconTypeName.Error]: '#d64646'
};

export const IconColorDisabledMapper = {
  [IconTypeName.Primary]: '#686d72',
  [IconTypeName.Inverted]: '#dbdbdb',
  [IconTypeName.Link]: '#c3d7fa',
  [IconTypeName.Success]: '#4bb678',
  [IconTypeName.Warning]: '#efa842',
  [IconTypeName.Error]: '#d64646'
};
