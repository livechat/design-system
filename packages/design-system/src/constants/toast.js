export const HORIZONTAL_POSITION = ['left', 'center', 'right'];

export const VERTICAL_POSITION = ['top', 'bottom'];

export const ANIMATION_TIME = 200;

export const INFO = 'info';
export const SUCCESS = 'success';
export const WARNING = 'warning';
export const ERROR = 'error';
export const NOTIFICATION = 'notification';

export const VARIANTS = [SUCCESS, WARNING, ERROR, INFO, NOTIFICATION];
