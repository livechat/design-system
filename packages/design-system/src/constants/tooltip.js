export const DIRECTION = {
  Top: 'top',
  Bottom: 'bottom',
  Left: 'left',
  Right: 'right'
};

export const ALIGMENT = {
  Top: 'top',
  Center: 'center',
  Bottom: 'bottom',
  Left: 'left',
  Right: 'right'
};

export const ARROW_POSITION = {
  Top: 'top',
  Bottom: 'bottom',
  Left: 'left',
  Right: 'right'
};

export const TOOLTIP_MARGIN = 4;
