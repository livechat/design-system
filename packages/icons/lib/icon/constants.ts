import { icons } from './icons';
import { IconSize, MaterialIcon, TablerIcon } from './types';

export const BASE_CLASS = 'icon';

export const IconSizeMap: Record<IconSize, { width: number; height: number }> =
  {
    xsmall: {
      width: 12,
      height: 12,
    },
    small: {
      width: 16,
      height: 16,
    },
    medium: {
      width: 20,
      height: 20,
    },
    large: {
      width: 24,
      height: 24,
    },
    xlarge: {
      width: 32,
      height: 32,
    },
  };
export const TablerIconsList = Object.keys(icons.tabler) as Array<TablerIcon>;
export const MaterialIconsList = Object.keys(
  icons.material
) as Array<MaterialIcon>;
