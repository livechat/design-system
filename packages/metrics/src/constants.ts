export enum ApplicationID {
  AGENT_APP = 'AGENT_APP',
  ACCOUNTS = 'ACCOUNTS',
  HELPDESK = 'HELPDESK',
}

export const FLAGMAN_DS_METRICS_PATH = '/webhooks/metrics/design-system';
