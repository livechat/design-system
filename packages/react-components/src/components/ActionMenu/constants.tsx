import * as React from 'react';

import { Icon } from '@livechat/design-system-icons';

import noop from '../../utils/noop';

import { ActionMenuItem } from './ActionMenuItem';

export const exampleOptions = [
  {
    key: 'one',
    element: (
      <ActionMenuItem leftNode={<Icon set="tabler" name="ContentCopy" />}>
        Copy
      </ActionMenuItem>
    ),
    onClick: noop,
  },
  {
    key: 'two',
    element: (
      <ActionMenuItem leftNode={<Icon set="tabler" name="MoveTo" />}>
        Move to...
      </ActionMenuItem>
    ),
    onClick: noop,
  },
  {
    key: 'three',
    element: (
      <ActionMenuItem leftNode={<Icon set="tabler" name="Edit" />}>
        Edit
      </ActionMenuItem>
    ),
    onClick: noop,
  },
  {
    key: 'four',
    element: (
      <ActionMenuItem leftNode={<Icon set="tabler" name="Block" />}>
        Block
      </ActionMenuItem>
    ),
    disabled: true,
    onClick: noop,
  },
  {
    key: 'five',
    element: (
      <ActionMenuItem
        kind="warning"
        leftNode={<Icon set="tabler" name="Delete" />}
      >
        Delete item
      </ActionMenuItem>
    ),
    onClick: noop,
    withDivider: true,
  },
];
