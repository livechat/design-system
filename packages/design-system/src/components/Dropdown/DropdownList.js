import * as React from 'react';
import * as PropTypes from 'prop-types';
import styles from './style.scss';
import getMergedClassNames from '../../utils/getMergedClassNames';
import DropdownListItem from './DropdownListItem';
import { KeyCodes } from '../../constants/keyCodes';
import findNextFocusableItem from '../../helpers/find-next-focusable-item';
import { getFirstFocusableItemId } from './helpers';

const baseClass = 'dropdown';

class DropdownList extends React.PureComponent {
  constructor(props) {
    super(props);

    this.state = {
      focusedElement:
        this.props.autoFocusedItemId || getFirstFocusableItemId(props.items),
      itemsCount: props.items.length
    };
  }

  static getDerivedStateFromProps(props, state) {
    if (
      props.autoFocusOnItemsCountChange &&
      props.items.length !== state.itemsCount
    ) {
      return {
        focusedElement: getFirstFocusableItemId(props.items),
        itemsCount: props.items.length
      };
    }
    return null;
  }

  componentDidMount() {
    if (this.props.keyboardEventsEnabled) {
      document.addEventListener('keydown', this.onKeydown);
    }

    if (this.props.autoFocusedItemId) {
      this.scrollItems();
    }
  }

  componentDidUpdate(prevProps) {
    const eventsEnabled =
      !prevProps.keyboardEventsEnabled && this.props.keyboardEventsEnabled;
    const eventsDisabled =
      prevProps.keyboardEventsEnabled && !this.props.keyboardEventsEnabled;

    if (eventsEnabled) {
      document.addEventListener('keydown', this.onKeydown);
    } else if (eventsDisabled) {
      document.removeEventListener('keydown', this.onKeydown);
    }
  }

  componentWillUnmount() {
    if (this.hoverEnablerTimeout) {
      clearTimeout(this.hoverEnablerTimeout);
    }
    document.removeEventListener('keydown', this.onKeydown);
  }

  onKeydown = event => {
    const { keyCode } = event;

    if (keyCode === KeyCodes.arrowDown || keyCode === KeyCodes.arrowUp) {
      this.handleArrowKeyUse(event);
    }

    if (this.isItemSelectKeyCode(keyCode)) {
      event.preventDefault();
      this.handleSelectKeyUse(event);
    }
  };

  getFocusedItemCallback = itemKey => {
    if (!this.hoverCallbacks[itemKey]) {
      this.hoverCallbacks[itemKey] = () => {
        if (!this.isHoverDisabled) {
          this.changeFocusedElement(itemKey);
        }
      };
    }

    return this.hoverCallbacks[itemKey];
  };

  handleSelectKeyUse = event => {
    const { focusedElement } = this.state;

    if (focusedElement !== null) {
      const selectedItem = this.props.items.find(
        item => item.itemId === focusedElement
      );

      if (selectedItem && selectedItem.onItemSelect) {
        selectedItem.onItemSelect(selectedItem.itemId, event);
      }
    }
  };

  handleArrowKeyUse = event => {
    event.preventDefault();
    const { keyCode } = event;
    const { items } = this.props;

    const nextItem = findNextFocusableItem(
      items,
      this.state.focusedElement,
      keyCode
    );

    if (nextItem) {
      this.changeFocusedElement(nextItem.itemId);
      this.scrollItems();
    }
  };

  handleListScroll = event => {
    event.preventDefault();
    if (this.props.onScroll) {
      this.props.onScroll(event);
    }
    this.enableHoverOnItems(150);
  };

  changeFocusedElement = id => {
    this.setState(
      {
        focusedElement: id
      },
      () => {
        const focusedItem = this.props.items.find(item => item.itemId === id);

        if (focusedItem && focusedItem.onItemFocus) {
          focusedItem.onItemFocus(focusedItem.itemdId);
        }
      }
    );
  };

  isItemSelectKeyCode = keyCode => {
    const { itemSelectKeyCodes } = this.props;

    if (itemSelectKeyCodes && itemSelectKeyCodes.includes(keyCode)) {
      return true;
    }

    return false;
  };

  scrollItems = () => {
    if (!this.listRef.current) {
      return;
    }
    const focusedElement = this.listRef.current.querySelector(
      `.lc-${baseClass}__list-item--focused`
    );

    if (focusedElement) {
      this.isHoverDisabled = true;
      const {
        height: ulHeight,
        top: ulTop
      } = this.listRef.current.getBoundingClientRect();
      const {
        height: itemHeigth,
        top: itemTop
      } = focusedElement.getBoundingClientRect();
      const relativeTop = itemTop + itemHeigth - ulTop;
      const itemOfsetTop = focusedElement.offsetTop;

      if (relativeTop > ulHeight) {
        this.listRef.current.scrollTop = itemOfsetTop - ulHeight + itemHeigth;
      } else if (itemTop < ulTop) {
        this.listRef.current.scrollTop =
          itemOfsetTop - (itemOfsetTop % itemHeigth);
      }
    }

    this.enableHoverOnItems(150);
  };

  enableHoverOnItems(delayInMs) {
    if (this.hoverEnablerTimeout) {
      clearTimeout(this.hoverEnablerTimeout);
    }

    if (delayInMs) {
      this.hoverEnablerTimeout = setTimeout(() => {
        this.isHoverDisabled = false;
      }, delayInMs);
    } else {
      this.isHoverDisabled = false;
    }
  }

  isHoverDisabled = false;
  hoverEnablerTimeout = null;

  hoverCallbacks = [];

  listRef = React.createRef();

  render() {
    const {
      className,
      items,
      getItemBody,
      itemSelectKeyCodes,
      autoFocusOnItemsCountChange,
      keyboardEventsEnabled,
      ...restProps
    } = this.props;

    const mergedClassNames = getMergedClassNames(
      styles[`${baseClass}__list`],
      className
    );

    return (
      <ul
        className={mergedClassNames}
        tabIndex={0}
        onScroll={this.handleListScroll}
        ref={this.listRef}
        {...restProps}
      >
        {items.map(
          ({ content, itemId, props, onItemFocus, ...itemRestProps }) => {
            const itemProps = {
              ...itemRestProps,
              itemId,
              isFocused: this.state.focusedElement === itemId,
              onMouseOverItem: this.getFocusedItemCallback(itemId)
            };

            if (this.props.getItemBody) {
              return this.props.getItemBody({
                ...itemProps,
                props: props || {},
                onItemFocus,
                content
              });
            }

            return (
              <DropdownListItem key={itemId} {...itemProps}>
                {content}
              </DropdownListItem>
            );
          }
        )}
      </ul>
    );
  }
}

DropdownList.propTypes = {
  /**
   * Specify which item should be focused after dropdown open
   */
  autoFocusedItemId: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  autoFocusOnItemsCountChange: PropTypes.bool,
  className: PropTypes.string,
  /**
   * use this property to enable/disable keyboard events of DropdownList
   */
  keyboardEventsEnabled: PropTypes.bool,
  items: PropTypes.arrayOf(
    PropTypes.shape({
      className: PropTypes.string,
      itemId: PropTypes.oneOfType([PropTypes.string, PropTypes.number])
        .isRequired,
      content: PropTypes.node,
      divider: PropTypes.bool,
      icon: PropTypes.node,
      onItemSelect: PropTypes.func,
      onItemFocus: PropTypes.func,
      isDisabled: PropTypes.bool,
      isSelected: PropTypes.bool,
      props: PropTypes.object
    })
  ).isRequired,
  getItemBody: PropTypes.func,
  onScroll: PropTypes.func,
  /**
   * you can specify which key press should trigger list item select
   */
  itemSelectKeyCodes: PropTypes.arrayOf(PropTypes.number)
};

DropdownList.defaultProps = {
  keyboardEventsEnabled: true,
  itemSelectKeyCodes: [KeyCodes.enter]
};

export default DropdownList;
