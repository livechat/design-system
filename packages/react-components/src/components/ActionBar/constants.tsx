import {
  AccountCircle,
  AddTemplate,
  Adjust,
  Apps,
  Calendar,
  Contact,
} from '@livechat/design-system-icons';

import { Icon } from '../Icon';

import { IActionBarOption } from './types';

export const getDefaultOptions = (
  onClickHandler: (key: string) => void
): IActionBarOption[] => [
  {
    key: 'one',
    element: <Icon source={AccountCircle} kind="primary" />,
    label: 'One test',
    showTooltip: true,
    onClick: () => onClickHandler('one'),
  },
  {
    key: 'two',
    element: <Icon source={AddTemplate} kind="primary" />,
    label: 'Two test',
    showTooltip: true,
    onClick: () => onClickHandler('two'),
  },
  {
    key: 'three',
    element: <Icon source={Adjust} kind="primary" />,
    label: 'Three test',
    showTooltip: true,
    onClick: () => onClickHandler('three'),
  },
  {
    key: 'four',
    element: <Icon source={Apps} kind="primary" />,
    label: 'Four test',
    showTooltip: true,
    onClick: () => onClickHandler('four'),
  },
  {
    key: 'five',
    element: <Icon source={Calendar} kind="primary" />,
    label: 'Five test',
    showTooltip: true,
    onClick: () => onClickHandler('five'),
  },
  {
    key: 'six',
    element: <Icon source={Contact} kind="primary" />,
    label: 'Six test',
    showTooltip: true,
    onClick: () => onClickHandler('six'),
  },
];
