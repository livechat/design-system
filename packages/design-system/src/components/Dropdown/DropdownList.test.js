import * as React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import { shallow } from 'enzyme';
import AlertCircleIcon from 'react-material-icon-svg/dist/AlertCircleIcon';
import DropdownList from './DropdownList';
import { KeyCodes } from '../../constants/keyCodes';

const generateItems = (length = 10) =>
  Array.from(new Array(length), (_value, index) => ({
    itemId: index + 1,
    divider: index === 2,
    content: `Item ${index + 1}`,
    icon: <AlertCircleIcon height={16} width={16} fill="#4384f5" />,
    isSelected: index === 2,
    isDisabled: index === 0,
    onItemSelect: jest.fn()
  }));

describe('Archives | Components | FiltersMenu', () => {
  let items;

  beforeEach(() => {
    items = generateItems(4);
  });

  it('renders correctly', () => {
    const renderer = new ShallowRenderer();
    const component = renderer.render(<DropdownList items={items} />);

    expect(component).toMatchSnapshot();
  });
});
