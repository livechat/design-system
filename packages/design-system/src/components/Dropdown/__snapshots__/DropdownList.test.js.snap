// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Archives | Components | FiltersMenu renders correctly 1`] = `
<ul
  className="dropdown__list"
  onScroll={[Function]}
  tabIndex={0}
>
  <DropdownListItem
    divider={false}
    icon={
      <AlertCircleIcon
        fill="#4384f5"
        height={16}
        width={16}
      />
    }
    isDisabled={true}
    isFocused={false}
    isSelected={false}
    itemId={1}
    onItemSelect={[MockFunction]}
    onMouseOverItem={[Function]}
  >
    Item 1
  </DropdownListItem>
  <DropdownListItem
    divider={false}
    icon={
      <AlertCircleIcon
        fill="#4384f5"
        height={16}
        width={16}
      />
    }
    isDisabled={false}
    isFocused={false}
    isSelected={false}
    itemId={2}
    onItemSelect={[MockFunction]}
    onMouseOverItem={[Function]}
  >
    Item 2
  </DropdownListItem>
  <DropdownListItem
    divider={true}
    icon={
      <AlertCircleIcon
        fill="#4384f5"
        height={16}
        width={16}
      />
    }
    isDisabled={false}
    isFocused={false}
    isSelected={true}
    itemId={3}
    onItemSelect={[MockFunction]}
    onMouseOverItem={[Function]}
  >
    Item 3
  </DropdownListItem>
  <DropdownListItem
    divider={false}
    icon={
      <AlertCircleIcon
        fill="#4384f5"
        height={16}
        width={16}
      />
    }
    isDisabled={false}
    isFocused={false}
    isSelected={false}
    itemId={4}
    onItemSelect={[MockFunction]}
    onMouseOverItem={[Function]}
  >
    Item 4
  </DropdownListItem>
</ul>
`;
